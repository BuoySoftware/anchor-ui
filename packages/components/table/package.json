{
  "name": "@buoysoftware/anchor-table",
  "version": "0.31.1",
  "main": "src/index.ts",
  "license": "MIT",
  "dependencies": {
    "@buoysoftware/anchor-button": "^0.31.1",
    "@buoysoftware/anchor-layout": "^0.31.1",
    "@buoysoftware/anchor-typography": "^0.31.1",
    "lodash": "^4.17.21",
    "react-i18next": "^12.1.4"
  },
  "scripts": {
    "build": "tsc -outDir dist",
    "compile": "tsc --noEmit",
    "postpack": "clean-package restore",
    "prepack": "clean-package",
    "test": "jest"
  },
  "devDependencies": {
    "jest": "^29.3.1",
    "ts-jest": "^29.0.4",
    "ts-node": "^10.9.1"
  }
}
