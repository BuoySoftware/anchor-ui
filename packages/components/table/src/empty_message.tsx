import { useTranslation } from "react-i18next";
import { Body } from "@buoysoftware/anchor-typography";
import { useTable, DEFAULT_TABLE_NAMESPACE } from "./use_table";

export const EmptyMessage: React.FC = (): React.ReactElement => {
  const { tNamespace, tableName } = useTable();
  const { t } = useTranslation();

  const translationLookup = [
    `${tNamespace}:${tableName}.empty_message`,
    `${tNamespace}:tables.${tableName}.empty_message`,
    `${DEFAULT_TABLE_NAMESPACE}:${tableName}.empty_message`,
    `${DEFAULT_TABLE_NAMESPACE}:empty_message`,
  ];

  return (
    <Body mt="xl" size="m" textAlign="center" data-testid="empty-table-state">
      {t(translationLookup)}
    </Body>
  );
};
