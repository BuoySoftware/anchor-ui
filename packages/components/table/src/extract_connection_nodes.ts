export interface Edge<NodeData> {
  node: NodeData;
}
export interface Connection<NodeData> {
  edges: Edge<NodeData>[];
}

export function extractConnectionNodes<NodeData>(
  connection?: Connection<NodeData>
): NodeData[] {
  if (!connection) return [];

  return connection.edges.map(({ node }) => node);
}
