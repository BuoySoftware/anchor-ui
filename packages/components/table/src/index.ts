export * from "./empty_cell";
export * from "./table";
