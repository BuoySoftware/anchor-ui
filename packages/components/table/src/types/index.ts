import type { PropertyPath } from "lodash";
import type { UniqueIdentifier } from "@dnd-kit/core";
import { BodyProps } from "@buoysoftware/anchor-typography";

// Pagination utility types
export interface Edge<NodeData> {
  node: Record<NodeData>;
}

export interface Connection<NodeData> {
  edges: Edge<NodeData>[];
}

export interface ConnectionWithEdges<EdgeData> {
  edges: EdgeData[];
}

export interface PageInfo {
  startCursor: string | null;
  endCursor: string | null;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
  perPage: number;
}

export interface PageableConnection {
  pageInfo: PageInfo;
  totalCount: number;
}

export interface PaginationVariables {
  variables: {
    after?: string;
    before?: string;
    first?: number;
    last?: number;
  };
}

export type FetchMore = (
  fetchMoreOptions: PaginationVariables
) => Promise<unknown>;

export type PaginationData<RecordData> = (
  | Connection<RecordData>
  | ConnectionWithEdges<RecordData>
) &
  PageableConnection;

// Table component Types
export type CellProps = Partial<BodyProps> | undefined;

export type Record<RecordData> = RecordData & { id?: string };

export type RowAction<RecordData> = (rowData: RecordData) => void;

export type TableVariant = "default" | "condensed";

export interface TableCellConfig<RecordData> {
  cellProps?: (rowData: RecordData) => CellProps;
  dataKey: PropertyPath;
  name?: string;
  render?: (rowData: RecordData) => React.ReactNode;
}

export interface TableProps<RecordData> {
  TableCellComponent?: React.FC<TableCell<RecordData>>;
  TableRowComponent?: React.FC<TableRow<RecordData>>;
  cellConfigs: TableCellConfig<RecordData>[];
  fetchMore?: FetchMore;
  name: string;
  paginationData?: PaginationData<RecordData>;
  recordIdKey: PropertyPath;
  records: Record<RecordData>[];
  renderPlaceholder?: RenderPlaceholder<RecordData>;
  rowAction?: RowAction<RecordData>;
  tNamespace?: string;
}

export interface TableBodyProps<RecordData> {
  TableCellComponent?: React.FC<TableCell<RecordData>>;
  TableRowComponent?: React.FC<TableRow<RecordData>>;
  cellConfigs: TableCellConfig<RecordData>[];
  recordIdKey: PropertyPath;
  records: Record<RecordData>[];
  renderPlaceholder?: RenderPlaceholder<RecordData>;
  rowAction?: RowAction<RecordData>;
}

export type RenderPlaceholder<RecordData> = (
  rowData: RecordData
) => React.ReactElement | string;

export interface TableRow<RecordData> {
  TableCellComponent?: React.FC<TableCell<RecordData>>;
  cellConfigs: TableCellConfig<RecordData>[];
  record: Record<RecordData>;
  renderPlaceholder?: RenderPlaceholder<RecordData>;
  rowAction?: RowAction<RecordData>;
  rowId: string;
}

export interface TableCell<RecordData> {
  cellConfig: TableCellConfig<RecordData>;
  record: RecordData;
  renderPlaceholder?: RenderPlaceholder<RecordData>;
  testId: string;
  variant?: TableVariant;
}

// Sortable table types
export interface SortableRecord extends Object {
  id: UniqueIdentifier;
}

export interface SortableTableCellConfig<RecordData>
  extends TableCellConfig<RecordData> {
  sortHandle?: boolean;
}

export interface SortableTableRow<RecordData>
  extends Omit<TableRow<RecordData>, "cellConfigs"> {
  cellConfigs: SortableTableCellConfig<RecordData>[];
  sortHandle?: boolean;
}

export interface SortableTableCell<RecordData>
  extends Omit<TableCell<RecordData>, "cellConfig"> {
  cellConfig: SortableTableCellConfig<RecordData>;
  sortHandle?: boolean;
}

export interface SortableTableProps<RecordData>
  extends Omit<
    TableProps<RecordData>,
    "cellConfigs" | "TableCellComponent" | "TableRowComponent"
  > {
  cellConfigs: SortableTableCellConfig<RecordData>[];
  TableCellComponent?: React.FC<SortableTableCell<RecordData>>;
  TableRowComponent?: React.FC<SortableTableRow<RecordData>>;
  onSort?: (records: Record<RecordData>[]) => void;
}
