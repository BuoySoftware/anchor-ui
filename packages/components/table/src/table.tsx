import kebabCase from "lodash/kebabCase";

import { StyledTable, StyledTableProps } from "./styled_table";
import { Tbody } from "./tbody";
import { Thead } from "./thead";
import { TableProvider } from "./table_provider";
import {
  Connection,
  ConnectionWithEdges,
  PageableConnection,
  RenderPlaceholder,
  TbodyProps,
} from "./types";
import { DEFAULT_TABLE_NAMESPACE } from "./use_table";
import { EmptyMessage } from "./empty_message";
import { FetchMore, TableActions } from "./table_actions";

interface OwnProps<RecordData> {
  name: string;
  fetchMore?: FetchMore;
  paginationData?: (Connection<RecordData> | ConnectionWithEdges<RecordData>) &
    PageableConnection;
  renderPlaceholder?: RenderPlaceholder<RecordData>;
  tNamespace?: string;
}

export type TableProps<RecordData> = OwnProps<RecordData> &
  TbodyProps<RecordData> &
  Partial<StyledTableProps>;

export const Table = <RecordData,>({
  cellConfigs,
  fetchMore,
  name,
  paginationData,
  renderPlaceholder = () => "--",
  rowAction,
  recordIdKey,
  records,
  tNamespace = DEFAULT_TABLE_NAMESPACE,
}: TableProps<RecordData>): React.ReactElement => {
  const renderTableActions = fetchMore && paginationData && records.length > 0;

  return (
    <TableProvider value={{ name, tNamespace }}>
      <StyledTable data-testid={`${kebabCase(name)}-table`}>
        <Thead cellConfigs={cellConfigs} />
        <Tbody
          cellConfigs={cellConfigs}
          records={records}
          renderPlaceholder={renderPlaceholder}
          rowAction={rowAction}
          recordIdKey={recordIdKey}
        />
      </StyledTable>
      {records.length === 0 && <EmptyMessage />}
      {renderTableActions && (
        <TableActions data={paginationData} fetchMore={fetchMore} />
      )}
    </TableProvider>
  );
};
