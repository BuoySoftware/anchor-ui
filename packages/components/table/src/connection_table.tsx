import { Table, TableProps } from "./table";
import { Connection, PageableConnection } from "./types";
import { extractConnectionNodes } from "./extract_connection_nodes";

interface OwnProps<RecordData> {
  connection?: Connection<RecordData> & PageableConnection;
}

type ConnectionTableProps<RecordData> = OwnProps<RecordData> &
  Omit<TableProps<RecordData>, "records" | "paginationData">;

export const ConnectionTable = <RecordData,>({
  connection,
  ...tableBuilderProps
}: ConnectionTableProps<RecordData>): React.ReactElement => {
  const records = extractConnectionNodes(connection);

  return (
    <Table
      paginationData={connection}
      records={records}
      {...tableBuilderProps}
    />
  );
};
