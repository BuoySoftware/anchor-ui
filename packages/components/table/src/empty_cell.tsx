import { TableCellConfig } from "./types";

export const EMPTY_CELL: TableCellConfig<any> = {
  dataKey: "",
  render: () => <></>,
};
