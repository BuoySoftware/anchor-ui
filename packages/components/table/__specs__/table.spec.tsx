import i18n from "i18next";
import { initReactI18next } from "react-i18next";
import { render as baseRender, RenderResult } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { theme } from "@buoysoftware/anchor-theme";

import { extractConnectionNodes } from "../src/extract_connection_nodes";
import { Connection, PageableConnection, TableCellConfig } from "../src/types";
import { Table } from "../src/table";

interface Donation {
  __typename: string;
  id: string;
  status: string | null;
}

const records: Donation[] = [
  { __typename: "Donation", id: "1", status: "Good" },
  { __typename: "Donation", id: "2", status: "Bad" },
  { __typename: "Donation", id: "3", status: "Ugly" },
  { __typename: "Donation", id: "4", status: "Other" },
];

const cellConfigs: TableCellConfig<Donation>[] = [
  { dataKey: "id" },
  { dataKey: "status" },
];

const connection: Connection<Donation> & PageableConnection = {
  edges: [
    {
      node: {
        __typename: "Donation",
        id: "1",
        status: "Good",
      },
    },
    {
      node: {
        __typename: "Donation",
        id: "2",
        status: "Bad",
      },
    },
    {
      node: {
        __typename: "Donation",
        id: "3",
        status: "Ugly",
      },
    },
    {
      node: {
        __typename: "Donation",
        id: "4",
        status: "Other",
      },
    },
  ],
  pageInfo: {
    endCursor: null,
    startCursor: null,
    hasPreviousPage: false,
    hasNextPage: true,
    perPage: 2,
  },
  totalCount: 4,
};

describe("<Table />", () => {
  const render = (ui: React.ReactElement): RenderResult => {
    return baseRender(<ThemeProvider theme={theme}>{ui}</ThemeProvider>);
  };

  const tables = {
    empty_message: "Default empty message",
    test_table_donations_empty: {
      empty_message: "Nothing to see here",
    },
  };

  const tablesCustom = {
    tables: {
      test_table_donations_empty: {
        empty_message: "Namespaced custom message",
      },
    },
  };

  i18n.use(initReactI18next).init({
    lng: "en",
    ns: ["tables"],
    resources: {
      en: {
        tables,
        tablesCustom,
      },
    },
  });

  it("renders the table correctly", () => {
    const component = render(
      <Table
        cellConfigs={cellConfigs}
        name="test_table_donations"
        records={records}
        recordIdKey="id"
      />
    );

    expect(component.asFragment()).toMatchSnapshot();
  });

  context("data can be paginated", () => {
    it("renders the table and table actions correctly", () => {
      const records = extractConnectionNodes(connection);

      const component = render(
        <Table
          cellConfigs={cellConfigs}
          fetchMore={jest.fn()}
          name="test_table_donations"
          paginationData={connection}
          records={records}
          recordIdKey="id"
        />
      );

      expect(component.getByTestId("table-actions")).toBeInTheDocument();
    });
  });

  context("some data is null", () => {
    it("renders the placeholder", () => {
      const records: Donation[] = [
        { __typename: "Donation", id: "1", status: "Good" },
        { __typename: "Donation", id: "2", status: "Bad" },
        { __typename: "Donation", id: "3", status: "Ugly" },
        { __typename: "Donation", id: "4", status: null },
      ];

      const component = render(
        <Table
          cellConfigs={cellConfigs}
          name="test_table_donations"
          records={records}
          recordIdKey="id"
        />
      );

      const nullCell = component.getByTestId("tr-4-td-status");

      expect(nullCell).toHaveTextContent("--");
    });
  });

  context("records are empty", () => {
    it("renders a default empty state message", () => {
      const records: Donation[] = [];
      const component = render(
        <Table
          cellConfigs={cellConfigs}
          name="test_table_donations"
          records={records}
          recordIdKey="id"
        />
      );

      expect(component.getByTestId("empty-table-state")).toHaveTextContent(
        "Default empty message"
      );
    });

    it("does not render the table actions", () => {
      const records: Donation[] = [];
      const component = render(
        <Table
          cellConfigs={cellConfigs}
          name="test_table_donations"
          records={records}
          recordIdKey="id"
        />
      );

      expect(component.queryByTestId("table-actions")).not.toBeInTheDocument();
    });

    context("there is a custom empty state message", () => {
      it("renders the custom message", () => {
        const records: Donation[] = [];
        const component = render(
          <Table
            cellConfigs={cellConfigs}
            name="test_table_donations_empty"
            records={records}
            recordIdKey="id"
          />
        );

        expect(component.getByTestId("empty-table-state")).toHaveTextContent(
          "Nothing to see here"
        );
      });
    });

    context("a custom tNamespace is provided", () => {
      context("custom empty message exists", () => {
        it("renders the custom message from the namespace", () => {
          const records: Donation[] = [];
          const component = render(
            <Table
              cellConfigs={cellConfigs}
              name="test_table_donations_empty"
              records={records}
              recordIdKey="id"
              tNamespace="tablesCustom"
            />
          );

          expect(component.getByTestId("empty-table-state")).toHaveTextContent(
            "Namespaced custom message"
          );
        });
      });

      context("custom empty message does not exist", () => {
        it("renders the default fallback message", () => {
          const records: Donation[] = [];
          const component = render(
            <Table
              cellConfigs={cellConfigs}
              name="test_table_no_message"
              records={records}
              recordIdKey="id"
              tNamespace="tablesCustom"
            />
          );

          expect(component.getByTestId("empty-table-state")).toHaveTextContent(
            "Default empty message"
          );
        });
      });
    });

    it("does not render the table actions", () => {
      const records: Donation[] = [];
      const component = render(
        <Table
          cellConfigs={cellConfigs}
          name="test_table_donations"
          records={records}
          recordIdKey="id"
        />
      );

      expect(component.queryByTestId("table-actions")).not.toBeInTheDocument();
    });
  });
});
