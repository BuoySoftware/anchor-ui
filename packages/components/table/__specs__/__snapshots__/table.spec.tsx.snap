// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> renders the table correctly 1`] = `
<DocumentFragment>
  .c2 {
  box-sizing: border-box;
  color: #101112;
  width: s;
  height: s;
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
}

.c5 {
  box-sizing: border-box;
  color: #101112;
  width: m;
  height: m;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
}

.c3 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-size: 0.875rem;
  font-weight: 600;
  line-height: 1rem;
}

.c6 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-style: none;
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1rem;
}

.c0 {
  width: 100%;
  box-sizing: border-box;
}

.c4 tr {
  border-bottom: 1px solid #BABEC3;
}

.c1 {
  border-bottom: 1px solid #BABEC3;
}

.c1 th {
  overflow: hidden;
  text-align: left;
  text-overflow: ellipsis;
  vertical-align: bottom;
}

.c1 th:last-child {
  padding-right: 0;
}

.c1:first-child th {
  padding-top: 0;
}

<table
    class="c0"
    data-testid="test-table-donations-table"
  >
    <thead
      class="c1"
    >
      <tr>
        <th
          class="c2 c3"
          color="text.primary"
          font-size="font-size-100"
          font-weight="600"
        >
          test_table_donations.th.id
        </th>
        <th
          class="c2 c3"
          color="text.primary"
          font-size="font-size-100"
          font-weight="600"
        >
          test_table_donations.th.status
        </th>
      </tr>
    </thead>
    <tbody
      class="c4"
    >
      <tr
        data-testid="tr-1"
      >
        <td
          class="c5 c6"
          color="text.primary"
          data-testid="tr-1-td-id"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          1
        </td>
        <td
          class="c5 c6"
          color="text.primary"
          data-testid="tr-1-td-status"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          Good
        </td>
      </tr>
      <tr
        data-testid="tr-2"
      >
        <td
          class="c5 c6"
          color="text.primary"
          data-testid="tr-2-td-id"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          2
        </td>
        <td
          class="c5 c6"
          color="text.primary"
          data-testid="tr-2-td-status"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          Bad
        </td>
      </tr>
      <tr
        data-testid="tr-3"
      >
        <td
          class="c5 c6"
          color="text.primary"
          data-testid="tr-3-td-id"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          3
        </td>
        <td
          class="c5 c6"
          color="text.primary"
          data-testid="tr-3-td-status"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          Ugly
        </td>
      </tr>
      <tr
        data-testid="tr-4"
      >
        <td
          class="c5 c6"
          color="text.primary"
          data-testid="tr-4-td-id"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          4
        </td>
        <td
          class="c5 c6"
          color="text.primary"
          data-testid="tr-4-td-status"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          Other
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
